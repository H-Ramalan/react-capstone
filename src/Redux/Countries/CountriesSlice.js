import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { v4 as uuidv4 } from 'uuid';
import axios from 'axios';

const url = 'https://restcountries.com/v3.1/all';
const detailUrl = 'https://restcountries.com/v3.1/name/';

const initialState = {
  data: [],
  loading: false,
  error: '',
  country: null,
};

const fetchData = (data) => data.map((country) => ({
  countryId: uuidv4(),
  officialName: country.name.official,
  commonName: country.name.common,
  capital: country.capital,
  area: country.area,
  landlocked: country.landlocked,
  independent: country.independent,
  timeZone: country.timezones,
  population: country.population,
  language: country.languages,
  continents: country.continents[0],
  coatOfArms: country.coatOfArms.png,
  flag: country.flags.png,
}));

export const fetchCountries = createAsyncThunk(
  'countries/fetchCountries',
  async () => {
    try {
      const response = await axios.get(url);
      const { data } = response;
      return fetchData(data);
    } catch (error) {
      throw Error(error);
    }
  },
);

export const fetchCountry = createAsyncThunk(
  'countries/fetchCountry',
  async (country) => {
    try {
      const response = await axios.get(detailUrl + country);

      const { data } = response;
      return fetchData(data);
    } catch (error) {
      throw Error(error);
    }
  },
);

export const CountriesSlice = createSlice({
  name: 'countries',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCountries.pending, (state) => ({ ...state, loading: true }))
      .addCase(fetchCountries.fulfilled, (state, action) => ({
        ...state,
        loading: false,
        data: action.payload,
      }))
      .addCase(fetchCountries.rejected, (state, action) => ({
        ...state,
        loading: false,
        error: action.error.message,
      }))
      .addCase(fetchCountry.pending, (state) => ({ ...state, loading: true }))
      .addCase(fetchCountry.fulfilled, (state, action) => ({
        ...state,
        loading: false,
        country: action.payload[0],
      }))
      .addCase(fetchCountry.rejected, (state, action) => ({
        ...state,
        loading: false,
        error: action.error.message,
      }));
  },
});

export default CountriesSlice.reducer;
