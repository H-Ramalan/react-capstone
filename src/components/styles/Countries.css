* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.link {
  text-decoration: none;
}

.arrow {
  width: 7%;
  height: 10%;
  color: white;
}

.flag {
  display: flex;
  justify-content: space-between;
}

.flag-pic {
  width: 60%;
  height: 25%;
}

.people {
  text-align: end;
  color: white;
  font-size: 18px;
}
