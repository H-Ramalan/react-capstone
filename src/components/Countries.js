import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import './styles/Countries.css';

const Countries = ({ country }) => {
  const { officialName, flag } = country;
  return (
    <>
      <div className="countries">
        <NavLink className="link" to={`/details/${officialName}`}>
          <div className="flag">
            <img src={flag} alt="flag" />
          </div>
          <h2 className="country-name">{officialName}</h2>
        </NavLink>
      </div>
    </>
  );
};

export default Countries;

Countries.propTypes = {
  country: PropTypes.shape({
    officialName: PropTypes.string.isRequired,
    flag: PropTypes.string.isRequired,
  }).isRequired,
};
