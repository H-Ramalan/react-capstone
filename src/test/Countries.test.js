// Countries.test.js
import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Countries from '../components/Countries';

describe('Countries Component', () => {
  const mockCountry = {
    officialName: 'Test Country',
    flag: 'test-flag-url',
    population: 1235458,
  };

  it('should render country name and flag', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Countries country={mockCountry} />
      </BrowserRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
