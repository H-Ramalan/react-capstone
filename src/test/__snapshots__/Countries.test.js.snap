// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries Component should render country name and flag 1`] = `
<DocumentFragment>
  <div
    class="countries"
  >
    <a
      class="link"
      href="/details/Test Country"
    >
      <div
        class="flag"
      >
        <img
          alt="flag"
          src="test-flag-url"
        />
      </div>
      <h2
        class="country-name"
      >
        Test Country
      </h2>
    </a>
  </div>
</DocumentFragment>
`;
